src/ejabberd_sm_cets.erl

```diff
@@ -0,0 +1,145 @@
+-module(ejabberd_sm_cets).
+
+-behavior(ejabberd_sm_backend).
+
+-include("mongoose.hrl").
+-include("session.hrl").
+
+-export([init/1,
+         get_sessions/0,
+         get_sessions/1,
+         get_sessions/2,
+         get_sessions/3,
+         create_session/4,
+         update_session/4,
+         delete_session/4,
+         cleanup/1,
+         total_count/0,
+         unique_count/0]).
+
+-define(TABLE, cets_session).
+
+-spec init(map()) -> any().
+init(_Opts) ->
+    cets:start(?TABLE, #{}),
+    cets_discovery:add_table(mongoose_cets_discovery, ?TABLE).
+
+-spec get_sessions() -> [ejabberd_sm:session()].
+get_sessions() ->
+    tuples_to_sessions(ets:tab2list(?TABLE)).
+
+-spec get_sessions(jid:lserver()) -> [ejabberd_sm:session()].
+get_sessions(Server) ->
+    %% This is not a full table scan. From the ETS docs:
+    %% For ordered_set a partially bound key will limit the traversal to only
+    %% scan a subset of the table based on term order.
+    %% A partially bound key is either a list or a tuple with
+    %% a prefix that is fully bound.
+    R = {{Server, '_', '_', '_'}, '_', '_'},
+    Xs = ets:select(?TABLE, [{R, [], ['$_']}]),
+    tuples_to_sessions(Xs).
+
+-spec get_sessions(jid:luser(), jid:lserver()) -> [ejabberd_sm:session()].
+get_sessions(User, Server) ->
+    R = {{Server, User, '_', '_'}, '_', '_'},
+    Xs = ets:select(?TABLE, [{R, [], ['$_']}]),
+    tuples_to_sessions(Xs).
+
+-spec get_sessions(jid:luser(), jid:lserver(), jid:lresource()) ->
+    [ejabberd_sm:session()].
+get_sessions(User, Server, Resource) ->
+    R = {{Server, User, Resource, '_'}, '_', '_'},
+    Xs = ets:select(?TABLE, [{R, [], ['$_']}]),
+    %% TODO these sessions should be deduplicated.
```

Comment:
Very similar in Mnesia: if user reconnects with the same resource name, we send kick to the old session, but that is an async operation, so we could have two records in the table.
Also, it could occur with cets and nesplits for sure, pretty rare though.

Reply:
Probably the best solution would be to write a property based test for this module (just test using the module API, without escalus and c2s logic).

We could find some interesting edge cases or at least we would figure out a set of properties the session API has.